quickwit-indexing/src/split_store/local_split_store.rs

```diff
@@ -0,0 +1,213 @@
+use std::{fs, io};
+
+use quickwit_common::split_file;
+use quickwit_storage::{LocalFileStorage, Storage, StorageErrorKind, StorageResult};
+use tracing::{error, warn};
+
+use super::IndexingSplitStoreParams;
+
+fn split_id_from_filename(dir_entry: &DirEntry) -> Option<String> {
+    if !dir_entry.path().is_file() {
+        return None;
+    }
+    let split_filename = dir_entry.file_name().into_string().ok()?;
+    split_filename
+        .strip_suffix(".split")
+        .map(ToString::to_string)
+}
+
+#[derive(Debug, Eq, PartialEq)]
+struct SizeInCache {
+    pub num_splits: usize,
+    pub size_in_bytes: usize,
+}
+
+pub struct LocalSplitStore {
+    /// The backing local storage.
+    local_storage: LocalFileStorage,
+    // The list of stored splits.
+    stored_splits: HashMap<String, usize>,
+    /// The parameters of the cache.
+    params: IndexingSplitStoreParams,
+}
+
+impl LocalSplitStore {
+    /// Try to open an existing local split directory.
+    ///
+    /// All files finishing by .split will be considered to be part of the directory.
+    pub fn open(
+        local_storage_root: PathBuf,
+        params: IndexingSplitStoreParams,
+    ) -> StorageResult<LocalSplitStore> {
+        let mut cache_entries = HashMap::new();
+        let mut total_size_in_bytes: usize = 0;
+        for dir_entry_result in fs::read_dir(&local_storage_root)? {
+            let dir_entry = dir_entry_result?;
+            if let Some(split_id) = split_id_from_filename(&dir_entry) {
+                let split_num_bytes = dir_entry.metadata()?.len() as usize;
+                total_size_in_bytes += split_num_bytes;
+                cache_entries.insert(split_id, split_num_bytes);
+            }
+        }
+
+        let local_storage = LocalFileStorage::from(local_storage_root);
+
+        if cache_entries.len() > params.max_num_splits {
+            return Err(StorageErrorKind::InternalError.with_error(anyhow::anyhow!(
+                "Initial number of files exceeds the maximum number of files allowed.",
+            )));
+        }
+
+        if total_size_in_bytes > params.max_num_bytes {
+            return Err(StorageErrorKind::InternalError.with_error(anyhow::anyhow!(
+                "Initial cache size exceeds the maximum size in bytes allowed.",
+            )));
+        }
+
+        Ok(LocalSplitStore {
+            local_storage,
+            stored_splits: cache_entries,
+            params,
+        })
+    }
+
+    #[cfg(test)]
+    pub fn inspect(&self) -> &HashMap<String, usize> {
+        &self.stored_splits
+    }
+
+    pub async fn remove_split(&mut self, split_id: &str) -> StorageResult<()> {
+        if !self.stored_splits.contains_key(split_id) {}
+        let split_path = PathBuf::from(quickwit_common::split_file(split_id));
+        if let Err(error) = self.local_storage.delete(&split_path).await {
+            warn!(file_path = %split_path.display(), error = %error, "Could not remove file from local storage.");
```

Comment:
hmmm we should probably return an error here too.

Reply:
I thought we wanted to be less strict on delete. but yes if we miss this delete, the file might never get a delete request anymore and even if we restart it will still be loaded